Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,634 @@
+          lines.pop_back();
+        }
+        auto& last_line = lines.back();
+        last_line = m_text_edit->fontMetrics().elidedText(last_line,
+          Qt::ElideRight, width() - get_border_size().width() -
+          get_padding_size().width() -
+          m_text_edit->fontMetrics().horizontalAdvance(ELLIPSES_CHAR));
+        if(is_elided && !last_line.endsWith(ELLIPSES_CHAR)) {
+          last_line.append(ELLIPSES_CHAR);
+        }
+        m_text_edit->blockSignals(true);
+        m_text_edit->setText(lines.join(""));
+        m_text_edit->blockSignals(false);
+      }
+    }
+    update_text_alignment();
+    update_document_line_height();
+    return;
+  } else if(m_text_edit->toPlainText() != m_model->get_current()) {
+    m_text_edit->setText(m_model->get_current());
+  }
+  update_text_alignment();
+  update_document_line_height();
+}
+
+void TextAreaBox::update_document_line_height() {
+  auto cursor_pos = m_text_edit->textCursor().position();
+  for(auto i = 0; i < m_text_edit->document()->blockCount(); ++i) {
+    auto block = m_text_edit->document()->findBlockByNumber(i);
+    if(block.isValid()) {
+      auto cursor = m_text_edit->textCursor();
+      cursor.setPosition(block.position());
+      m_text_edit->setTextCursor(cursor);
+      auto block_format = cursor.blockFormat();
+      block_format.setLineHeight(m_computed_line_height,
+        QTextBlockFormat::FixedHeight);
+      cursor.setBlockFormat(block_format);
+      m_text_edit->setTextCursor(cursor);
+    }
+  }
+  auto cursor = m_text_edit->textCursor();
+  cursor.setPosition(cursor_pos);
+  m_text_edit->setTextCursor(cursor);
+}
+
+void TextAreaBox::update_layout() {
+  update_text_edit_width();
+  m_stacked_widget->setMinimumSize(size() - get_border_size() -
+    get_padding_size());
+  m_stacked_widget->adjustSize();
+  update_display_text();
+  update_placeholder_text();
+}
+
+void TextAreaBox::update_line_height() {
+  m_computed_line_height =
+    static_cast<int>(static_cast<double>(m_text_edit->font().pixelSize()) *
+    *m_text_edit_styles.m_line_height);
+  m_placeholder->set_line_height(m_computed_line_height);
+  m_scroll_box->get_vertical_scroll_bar().set_line_size(
+    m_computed_line_height);
+  update_display_text();
+}
+
+void TextAreaBox::update_placeholder_text() {
+  if(is_placeholder_shown()) {
+    m_placeholder->set_text(m_placeholder_text);
+    m_placeholder->setFixedSize(size() - get_border_size() -
+      get_padding_size());
+    m_stacked_widget->adjustSize();
+    m_placeholder->show();
+  } else {
+    m_placeholder->hide();
+  }
+}
+
+void TextAreaBox::update_text_alignment() {
+  auto alignment = m_text_edit_styles.m_alignment.value_or(Qt::AlignLeft);
+  auto cursor_pos = m_text_edit->textCursor().position();
+  for(auto i = 0; i < m_text_edit->document()->blockCount(); ++i) {
+    auto block = m_text_edit->document()->findBlockByNumber(i);
+    if(block.isValid()) {
+      auto cursor = m_text_edit->textCursor();
```

Comment:
This looks very similar to some code written above. See if you can extract the common code.